mentat/session.py

```diff
@@ -116,24 +119,37 @@ def start(self) -> asyncio.Task[None]:
         async def run_main():
             try:
                 await self._main()
-                await self.stop()
-            except asyncio.CancelledError:
+            except (SessionExit, CancelledError):
                 pass
+            except (MentatError, UserError) as e:
+                self.stream.send(str(e), color="red")
             except Exception:
-                traceback.print_exc()
+                # All unhandled exceptions end up here
+                self.stream.send(
+                    f"Unhandled Exception: {traceback.format_exc()}", color="red"
+                )
+            finally:
+                await self.stop()
 
         setup_logging()
         self._main_task: Task[None] = asyncio.create_task(run_main())
         return self._main_task
 
     async def stop(self):
+        if self.stopped:
+            return
+        self.stopped = True
```

Comment:
We're setting the session as "stopped" here, but there's still stopping-logic that's running afterwards that can raise Exceptions (and therefore not complete the whole stop-sequence).

I think `self.stopped = True` should be set at the end of the `stop` method, and there should be some variable/lock that prevents `stop` from being called concurrently.

Reply:
What Exceptions can be raised? I'm not entirely sure what better way there is to do this; if we raise an Exception during shutdown, there's really no way to recover (since running stop again is just going to raise the Exception again most likely). I only added in the stopped variable to prevent this function from being called twice, not to indicate that the session is stopped. The reason the stop function is called twice is because we stop the Client, which calls to stop the Session again; I think I might fix this, since I think the Client should instead signal to the server that it needs to stop, and then the server will be in charge of stopping itself and telling the client to stop with the exit signal.

Reply:
> if we raise an Exception during shutdown, there's really no way to recover (since running stop again is just going to raise the Exception again most likely)

Yeah that makes sense. If an Exception is thrown during shutdown it should be up to the Server (which is managing the Session) on how to handle it. Now that I'm looking at it again, there's not really any way an Exception can be thrown anyways which is good.

How do we check if a Session is stopped though? There's a race condition of `self.stopped` being `True` while `self._exit_task` and `self._main_task` are cancelling (i.e. the Session is still stopping)? With how the Terminal Client handles Session shutdown (in `self._listen_for_client_exit`) this isn't a problem, however I think it makes more sense if we have a `self.stopping` (which replaces `self.stopped` here) and a `self.stopped` which is set at the end of the `self._stop` method (which can be set in the `finally` section of a try-catch block). Thoughts?